File: mobile/ios/ArmadaRemote/Invite.swift
import Foundation

enum RelayKind: String {
    case pair
    case op
}

struct RelayInvite: Equatable {
    var kind: RelayKind
    var relay: String
    var fleet: String
    var tokenOrSecret: String
}

enum RelayUriError: String, Error {
    case invalid
    case incomplete
    case insecure
}

enum RelayInviteParser {
    private static let fleetRe = try! NSRegularExpression(pattern: "^[a-z0-9-]{8,64}$")
    private static let hex64Re = try! NSRegularExpression(pattern: "^[a-f0-9]{64}$")

    static func origin(of relay: String) -> String? {
        guard let u = URL(string: relay), let scheme = u.scheme, let host = u.host else { return nil }
        let loopback = host == "127.0.0.1" || host == "localhost"
        let hostPort = u.port.map { "\(host):\($0)" } ?? host
        if scheme == "https" { return "\(scheme)://\(hostPort)" }
        if scheme == "http" && loopback { return "\(scheme)://\(hostPort)" }
        return nil
    }

    static func parse(_ input: String) -> Result<RelayInvite, RelayUriError> {
        guard let url = URL(string: input.trimmingCharacters(in: .whitespacesAndNewlines)) else {
            return .failure(.invalid)
        }
        guard url.scheme == "armada-relay" else { return .failure(.invalid) }
        let kindRaw = (url.host ?? "") + url.path
        let kind: RelayKind?
        switch kindRaw.replacingOccurrences(of: "/", with: "") {
        case "pair": kind = .pair
        case "op": kind = .op
        default: kind = nil
        }
        guard let kind else { return .failure(.invalid) }
        let items = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems ?? []
        func q(_ name: String) -> String {
            items.first(where: { $0.name == name })?.value?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        }
        let relayRaw = q("relay")
        let fleet = q("fleet")
        if relayRaw.isEmpty || fleet.isEmpty { return .failure(.incomplete) }
        guard let origin = origin(of: relayRaw) else { return .failure(.insecure) }
        if !matches(fleetRe, fleet) { return .failure(.incomplete) }
        let cred = kind == .pair ? q("secret") : q("token")
        if !matches(hex64Re, cred) { return .failure(.incomplete) }
        return .success(RelayInvite(kind: kind, relay: origin, fleet: fleet, tokenOrSecret: cred))
    }

    private static func matches(_ re: NSRegularExpression, _ s: String) -> Bool {
        re.firstMatch(in: s, range: NSRange(location: 0, length: (s as NSString).length)) != nil
    }
}

File: mobile/android/core/src/main/kotlin/app/armada/remote/Invite.kt
package app.armada.remote

private val FLEET_RE = Regex("^[a-z0-9-]{8,64}$")
private val HEX64 = Regex("^[a-f0-9]{64}$")

data class RelayInvite(
    val kind: String,
    val relay: String,
    val fleet: String,
    val cred: String,
)

sealed class InviteParse {
    data class Ok(val invite: RelayInvite) : InviteParse()
    data class Err(val code: String) : InviteParse()
}

object Invite {
    fun originOf(relay: String): String? {
        val u = runCatching { java.net.URI(relay) }.getOrNull() ?: return null
        val host = u.host ?: return null
        val scheme = u.scheme ?: return null
        val loopback = host == "127.0.0.1" || host == "localhost" || host == "10.0.2.2"
        val hostPort = if (u.port > 0) "$host:${u.port}" else host
        if (scheme == "https") return "$scheme://$hostPort"
        if (scheme == "http" && loopback) return "$scheme://$hostPort"
        return null
    }

    fun parse(input: String): InviteParse {
        val raw = input.trim()
        val url = runCatching { java.net.URI(raw) }.getOrNull() ?: return InviteParse.Err("invalid")
        if (url.scheme != "armada-relay") return InviteParse.Err("invalid")
        val kindRaw = ((url.host ?: "") + (url.path ?: "")).replace("/", "")
        val kind = when (kindRaw) {
            "pair" -> "pair"
            "op" -> "op"
            else -> return InviteParse.Err("invalid")
        }
        val q = query(url)
        val relayRaw = q["relay"]?.trim().orEmpty()
        val fleet = q["fleet"]?.trim().orEmpty()
        if (relayRaw.isEmpty() || fleet.isEmpty()) return InviteParse.Err("incomplete")
        val origin = originOf(relayRaw) ?: return InviteParse.Err("insecure")
        if (!FLEET_RE.matches(fleet)) return InviteParse.Err("incomplete")
        val cred = if (kind == "pair") {
            listOf(q["code"], q["secret"]).map { it?.trim().orEmpty() }.firstOrNull { HEX64.matches(it) }.orEmpty()
        } else {
            q["token"]?.trim().orEmpty()
        }
        if (!HEX64.matches(cred)) return InviteParse.Err("incomplete")
        return InviteParse.Ok(RelayInvite(kind, origin, fleet, cred))
    }

    private fun query(uri: java.net.URI): Map<String, String> {
        val raw = uri.rawQuery ?: return emptyMap()
        return raw.split("&").mapNotNull { part ->
            val i = part.indexOf("=")
            if (i <= 0) null
            else java.net.URLDecoder.decode(part.substring(0, i), Charsets.UTF_8) to
                java.net.URLDecoder.decode(part.substring(i + 1), Charsets.UTF_8)
        }.toMap()
    }
}
